Remove unecessary dependencies

MarkEWaite opened the pull request:
## Remove unnecessary dependencies

Tests complete without a dependency on workflow-cps-global-lib.

PR https://github.com/jenkinsci/workflow-cps-global-lib-plugin/pull/172 has deprecated workflow-cps-global-lib in
favor of pipeline-groovy-lib-plugin.  Remove the dependency because it is not required.

Adjust other dependencies to either remove them or make them `test` scoped.  Compiled and tested with Java 8 and Java 11.

MarkEWaite committed:
Remove unnecessary workflow-cps-global-lib dependency

basil reviewed (approved):
Thanks for the PR!

Optional feedback for your consideration: Thinking more deeply about why this unused dependency was present in the first place, I suspect it was only added to deal with upper bounds errors (which is now irrelevant with the use of the plugin BOM). Running `mvn dependency:analyze` shows other unused dependencies as well, most likely originally added for the same (now irrelevant) reason. If desired, the scope of this PR could be expanded to remove these as well. The benefit of this would be to slim down the dependency tree of this plugin, reducing future maintenance burden. This is not a blocking comment, since this PR is an improvement over the status quo and is welcome as-is. The following additional unused dependencies were identified by `mvn dependency:analyze`:

- `junit`
- `script-security`
- `ssh-credentials`
- `structs`
- `workflow-api`
- `workflow-cps-global-lib`
- `workflow-durable-task-step`
- `workflow-scm-step`
- `workflow-support`

MarkEWaite committed:
Remove unused junit plugin dependency

MarkEWaite committed:
Remove unused structs plugin dependency

Structs plugin as provider of the `@Symbol` annotation has been
replaced by another plugin.

MarkEWaite committed:
Remove unused script security plugin depenedency

MarkEWaite committed:
Exclude commons-lang3 from CasC plugin

The test scoped mockserver and mockclient currently provide a newer
version of commons-lang3 than the version provided with configuration
as code plugin.  Since configuration as code plugin is test scoped as
well, it seems safe to exclude commons-lang3 that is provided by
configuration as code plugin.

MarkEWaite committed:
Several workflow plugins are only test dependencies

No need to complicate the compile time dependencies when they are not
required for compilation.

MarkEWaite commented:
I assume there a few alternatives to make the CI job pass.  

Options I see:

* Discard the changes that reduce the dependencies because they also seem to increase the spotbugs warnings.  Return to a single commit that removes a single dependency and rely on someone else to make any other improvements.  I believe the other improvements in this pull request are correct and complete, but the spotbugs failures due to 5 new spotbugs warnings are blocking the pull request
* Close the pull request and accept that I'll remove the gitlab plugin from my installation, accepting that it reduces exploratory testing in my environment
* Suppress spotbugs warnings execution on the plugin completely
* Spend hours suppressing newly introduced spotbugs warnings

Recommendations @basil?

basil commented:
I chose a variation of 3, since SpotBugs was basically ignored in this plugin anyway. In https://github.com/jenkinsci/gitlab-plugin/pull/1293 I changed `<spotbugs.failOnError>false</spotbugs.failOnError>` to `<spotbugs.threshold>High</spotbugs.threshold>` so that the "Quality Gate" won't interfere with unrelated PRs like yours. Thanks for the PR! Released in [1.5.33](https://github.com/jenkinsci/gitlab-plugin/releases/tag/gitlab-plugin-1.5.33).

jglick commented on pom.xml:
```diff
@@ -87,25 +87,14 @@
-    <dependency>
-      <groupId>org.jenkins-ci.plugins</groupId>
-      <artifactId>script-security</artifactId>
+      <scope>test</scope>
```
Unlikely you need to specify transitive `test`-scoped dependencies like this either.
